README.md
# Stripe Ruby Library [![Build Status](https://travis-ci.org/stripe/stripe-ruby.svg?branch=master)](https://travis-ci.org/stripe/stripe-ruby)

The Stripe Ruby library provides convenient access to the Stripe API from
applications written in the Ruby language. It includes a pre-defined set of
classes for API resources that initialize themselves dynamically from API
responses which makes it compatible with a wide range of versions of the Stripe
API.

The library also provides other features. For example:

* Easy configuration path for fast setup and use.
* Helpers for pagination.
* Tracking of "fresh" values in API resources so that partial updates can be
  executed.
* Built-in mechanisms for the serialization of parameters according to the
  expectations of Stripe's API.

## Documentation

See the [Ruby API docs](https://stripe.com/docs/api/ruby#intro).

## Installation

You don't need this source code unless you want to modify the gem. If you just
want to use the package, just run:

    gem install stripe

If you want to build the gem from source:

    gem build stripe.gemspec

### Requirements

* Ruby 2.0+.

### Bundler

If you are installing via bundler, you should be sure to use the https rubygems
source in your Gemfile, as any gems fetched over http could potentially be
compromised in transit and alter the code of gems fetched securely over https:

``` ruby
source 'https://rubygems.org'

gem 'rails'
gem 'stripe'
```

## Usage

The library needs to be configured with your account's secret key which is
available in your [Stripe Dashboard][api-keys]. Set `Stripe.api_key` to its
value:

``` ruby
require "stripe"
Stripe.api_key = "sk_test_..."

# list charges
Stripe::Charge.list()

# retrieve single charge
Stripe::Charge.retrieve(
  "ch_18atAXCdGbJFKhCuBAa4532Z",
)
```

### Per-request Configuration

For apps that need to use multiple keys during the lifetime of a process, like
one that uses [Stripe Connect][connect], it's also possible to set a
per-request key and/or account:

``` ruby
require "stripe"

Stripe::Charge.list(
  {},
  :api_key => "sk_test_...",
  :stripe_account => "acct_..."
)

Stripe::Charge.retrieve(
  "ch_18atAXCdGbJFKhCuBAa4532Z",
  :api_key => "sk_test_...",
  :stripe_account => "acct_..."
)
```

### Configuring a Client

While a default HTTP client is used by default, it's also possible to have the
library use any client supported by [Faraday][faraday] by initializing a
`Stripe::StripeClient` object and giving it a connection:

``` ruby
conn = Faraday.new
client = Stripe::StripeClient.new(conn)
charge, resp = client.request do
  Stripe::Charge.retrieve(
    "ch_18atAXCdGbJFKhCuBAa4532Z",
  )
end
puts resp.request_id
```

### Configuring CA Bundles

By default, the library will use its own internal bundle of known CA
certificates, but it's possible to configure your own:

    Stripe.ca_bundle_path = "path/to/ca/bundle"

### Configuring Automatic Retries

The library can be configured to automatically retry requests that fail due to
an intermittent network problem:

    Stripe.max_network_retries = 2

[Idempotency keys][idempotency-keys] are added to requests to guarantee that
retries are safe.

### Configuring Timeouts

Open and read timeouts are configurable:

```java
Stripe.open_timeout = 30 // in seconds
Stripe.read_timeout = 80
```

Please take care to set conservative read timeouts. Some API requests can take
some time, and a short timeout increases the likelihood of a problem within our
servers.

### Logging

The library can be configured to emit logging that will give you better insight
into what it's doing. The `info` logging level is usually most appropriate for
production use, but `debug` is also available for more verbosity.

There are a few options for enabling it:

1. Set the environment variable `STRIPE_LOG` to the value `debug` or `info`:
   ```
   $ export STRIPE_LOG=info
   ```

2. Set `Stripe.log_level`:
   ``` ruby
   Stripe.log_level = Stripe::LEVEL_INFO
   ```

### Writing a Plugin

If you're writing a plugin that uses the library, we'd appreciate it if you
identified using `#set_app_info`:

    Stripe.set_app_info("MyAwesomePlugin", version: "1.2.34", url: "https://myawesomeplugin.info");

This information is passed along when the library makes calls to the Stripe
API.

## Development

The test suite depends on [stripe-mock], so make sure to fetch and run it from a
background terminal ([stripe-mock's README][stripe-mock] also contains
instructions for installing via Homebrew and other methods):

    go get -u github.com/stripe/stripe-mock
    stripe-mock

Run all tests:

    bundle exec rake test

Run a single test suite:

    bundle exec ruby -Ilib/ test/stripe/util_test.rb

Run a single test:

    bundle exec ruby -Ilib/ test/stripe/util_test.rb -n /should.convert.names.to.symbols/

Run the linter:

    bundle exec rake rubocop

Update bundled CA certificates from the [Mozilla cURL release][curl]:

    bundle exec rake update_certs

Update the bundled [stripe-mock] by editing the version number found in
`.travis.yml`.

[api-keys]: https://dashboard.stripe.com/account/apikeys
[connect]: https://stripe.com/connect
[curl]: http://curl.haxx.se/docs/caextract.html
[faraday]: https://github.com/lostisland/faraday
[idempotency-keys]: https://stripe.com/docs/api/ruby#idempotent_requests
[stripe-mock]: https://github.com/stripe/stripe-mock

<!--
# vim: set tw=79:
-->


History.txt
=== 3.5.1 2017-10-12

* [#591] Use thread-local `StripeClient` instances for thread safety

=== 3.5.0 2017-10-11

* [#589] Rename source `delete` to `detach` (and deprecate the former)

=== 3.4.1 2017-10-05

* [#586] Log query strings as well as form bodies with STRIPE_LOG
* [#588] Require minimum Faraday 0.10 for bug fix in parameter encoding

=== 3.4.0 2017-09-20

* Mark legacy Bitcoin API as deprecated, and remove corresponding tests
* Mark recipients API as deprecated, and remove recipient card tests

=== 3.3.2 2017-09-20

* Correct minimum required Ruby version in gemspec (it's 2.0.0)

=== 3.3.1 2017-08-18

* Only parse webhook payload after verification to decrease likelihood of
  attack

=== 3.3.0 2017-08-11

* Add support for standard library logger interface with `Stripe.logger`
* Error logs now go to stderr if using `Stripe.log_level`/`STRIPE_LOG`
* `Stripe.log_level`/`STRIPE_LOG` now support `Stipe::LEVEL_ERROR`

=== 3.2.0 2017-08-03

* Add logging for request retry account and `Stripe-Account` header

=== 3.1.0 2017-08-03

* Implement request logging with `Stripe.log_level` and `STRIPE_LOG`

=== 3.0.3 2017-07-28

* Revert `nil` to empty string coercion from 3.0.2
* Handle `invalid_client` OAuth error code
* Improve safety of error handling logic safer for unrecognized OAuth error
  codes

=== 3.0.2 2017-07-12 (yanked)

* Convert `nil` to empty string when serializing parameters (instead of
  opaquely dropping it) -- NOTE: this change has since been reverted

=== 3.0.1 2017-07-11

* Properties set with an API resource will now serialize that resource's ID if
  possible
* API resources will throw an ArgumentError on save if a property has been with
  an API resource that cannot be serialized

=== 3.0.0 2017-06-27

* `#pay` on invoice now takes params as well as opts

=== 2.12.0 2017-06-20

* Add support for ephemeral keys

=== 2.11.0 2017-05-26

* Warn when keys that look like opts are included as parameters

=== 2.10.0 2017-05-25

* Add support for account login links

=== 2.9.0 2017-05-18

* Support for OAuth operations in `Stripe::OAuth`

=== 2.8.0 2017-04-28

* Support for checking webhook signatures

=== 2.7.0 2017-04-26

* Add model `InvoiceLineItem`

=== 2.6.0 2017-04-26

* Add `OBJECT_NAME` constants to all API resources

=== 2.5.0 2017-04-24

* Make `opts` argument in `Util.convert_to_stripe_object` optional

=== 2.4.0 2017-04-18

* Add `Stripe.set_app_info` for use by plugin creators

=== 2.3.0 2017-04-14

* Add question mark accessor when assigning boolean value to undefined field

=== 2.2.1 2017-04-07

* Declare minimum required Faraday as 0.9

=== 2.2.0 2017-04-06

* Add support for payouts and recipient transfers

=== 2.1.0 2017-03-17

* Support for detaching sources from customers

=== 2.0.3 2017-03-16

* Fix marshalling of `StripeObjects` that have an embedded client

=== 2.0.2 2017-03-16

* Fix bad field reference when recovering from a JSON parsing problem

=== 2.0.1 2017-02-22

* Fix multipart parameter encoding to repair broken file uploads

=== 2.0.0 2017-02-14

* Drop support for Ruby 1.9
* Allow HTTP client that makes Stripe calls to be configured via Faraday
* Drop RestClient
* Switch to OpenAPI 2.0 spec and generated fixtures in test suite
* Switch to Webmock in test suite

=== 1.58.0 2017-01-19

* Remove erroneously added list methods for `Source` model

=== 1.57.1 2016-11-28

* Disallow sending protected fields along with API resource `.update`

=== 1.57.0 2016-11-21

* Add retrieve method for 3-D Secure resources

=== 1.56.2 2016-11-17

* Improve `StripeObject`'s `#to_s` to better handle how embedded objects are displayed

=== 1.56.1 2016-11-09

* Fix (fairly serious) memory like in `StripeObject`

=== 1.56.0 2016-10-24

* Add accessors for new fields added in `#update_attributes`
* Handle multi-plan subscriptions through new subscription items
* Handle 403 status codes from the API

=== 1.55.1 2016-10-24

Identical to 1.56.0 above. I incorrectly cut a patch-level release.

=== 1.55.0 2016-09-15

* Add support for Apple Pay domains

=== 1.54.0 2016-09-01

* Whitelist errors that should be retried; scope to known socket and HTTP errors

=== 1.53.0 2016-08-31

* Relax version constraint on rest-client (and by extension mime-types) for users on Ruby 2+

=== 1.52.0 2016-08-30

* Make sure `Subscription`'s `source` is saved with its parent

=== 1.51.1 2016-08-30

* Make sure `Account`'s `external_account` is saved with its parent

=== 1.51.0 2016-08-26

* Error when an array of maps is detected that cannot be accurately encoded
* Start using strings for header names instead of symbols for better clarity

=== 1.50.1 2016-08-25

* Fix encoding of arrays of maps where maps unequal sets of keys

=== 1.50.0 2016-08-15

* Allow sources to be created

=== 1.49.0 2016-07-28

* Add top-level `Source` model

=== 1.48.0 2016-07-12

* Add `ThreeDSecure` model for 3-D secure payments

=== 1.47.0 2016-07-11

* Allow rest-client version 2.0+ to be used with the gem

=== 1.46.0 2016-07-07

* Allow retry when a 409 conflict is encountered

=== 1.45.0 2016-07-07

* Do not send subresources when updating except when explicitly told to do so (see #433)

=== 1.44.0 2016-06-29

* Add `update` class method to all resources that can be updated

=== 1.43.1 2016-06-17

* Fix type of resource returned from `Order#return_order`

=== 1.43.0 2016-05-20

* Allow Relay orders to be returned and add associated types
* Support Alipay account retrieval and deletion

=== 1.42.0 2016-05-04

* Add support for the new /v1/subscriptions endpoint
  * Stripe::Subscription.retrieve
  * Stripe::Subscription.list
  * Stripe::Subscription.create
  * Stripe::Subscription.update
  * Stripe::Subscription.delete

=== 1.41.0 2016-04-13

* Add global `stripe_account` option that adds a `Stripe-Account` header to all requests

=== 1.40.0 2016-04-06

* Fix bug that omitted subresources from serialization

=== 1.39.0 2016-03-31

* Update CA cert bundle for compatibility with OpenSSL versions below 1.0.1


lib/stripe.rb
# Stripe Ruby bindings
# API spec at https://stripe.com/docs/api
require "cgi"
require "faraday"
require "json"
require "logger"
require "openssl"
require "rbconfig"
require "securerandom"
require "set"
require "socket"
require "uri"

# Version
require "stripe/version"

# API operations
require "stripe/api_operations/create"
require "stripe/api_operations/save"
require "stripe/api_operations/delete"
require "stripe/api_operations/list"
require "stripe/api_operations/request"
require "stripe/api_operations/nested_resource"

# API resource support classes
require "stripe/errors"
require "stripe/util"
require "stripe/stripe_client"
require "stripe/stripe_object"
require "stripe/stripe_response"
require "stripe/list_object"
require "stripe/api_resource"
require "stripe/singleton_api_resource"
require "stripe/webhook"

# Named API resources
require "stripe/account"
require "stripe/alipay_account"
require "stripe/apple_pay_domain"
require "stripe/application_fee"
require "stripe/application_fee_refund"
require "stripe/balance"
require "stripe/balance_transaction"
require "stripe/bank_account"
require "stripe/bitcoin_receiver"
require "stripe/bitcoin_transaction"
require "stripe/card"
require "stripe/charge"
require "stripe/country_spec"
require "stripe/coupon"
require "stripe/customer"
require "stripe/dispute"
require "stripe/ephemeral_key"
require "stripe/event"
require "stripe/file_upload"
require "stripe/invoice"
require "stripe/invoice_item"
require "stripe/invoice_line_item"
require "stripe/login_link"
require "stripe/order"
require "stripe/order_return"
require "stripe/payout"
require "stripe/plan"
require "stripe/product"
require "stripe/recipient"
require "stripe/recipient_transfer"
require "stripe/refund"
require "stripe/reversal"
require "stripe/sku"
require "stripe/source"
require "stripe/subscription"
require "stripe/subscription_item"
require "stripe/three_d_secure"
require "stripe/token"
require "stripe/transfer"

# OAuth
require "stripe/oauth"

module Stripe
  DEFAULT_CA_BUNDLE_PATH = File.dirname(__FILE__) + "/data/ca-certificates.crt"

  @app_info = nil

  @api_base = "https://api.stripe.com"
  @connect_base = "https://connect.stripe.com"
  @uploads_base = "https://uploads.stripe.com"

  @log_level = nil
  @logger = nil

  @max_network_retries = 0
  @max_network_retry_delay = 2
  @initial_network_retry_delay = 0.5

  @ca_bundle_path = DEFAULT_CA_BUNDLE_PATH
  @ca_store = nil
  @verify_ssl_certs = true

  @open_timeout = 30
  @read_timeout = 80

  class << self
    attr_accessor :stripe_account, :api_key, :api_base, :verify_ssl_certs, :api_version, :client_id, :connect_base, :uploads_base,
                  :open_timeout, :read_timeout

    attr_reader :max_network_retry_delay, :initial_network_retry_delay
  end

  # Gets the application for a plugin that's identified some. See
  # #set_app_info.
  def self.app_info
    @app_info
  end

  def self.app_info=(info)
    @app_info = info
  end

  # The location of a file containing a bundle of CA certificates. By default
  # the library will use an included bundle that can successfully validate
  # Stripe certificates.
  def self.ca_bundle_path
    @ca_bundle_path
  end

  def self.ca_bundle_path=(path)
    @ca_bundle_path = path

    # empty this field so a new store is initialized
    @ca_store = nil
  end

  # A certificate store initialized from the the bundle in #ca_bundle_path and
  # which is used to validate TLS on every request.
  #
  # This was added to the give the gem "pseudo thread safety" in that it seems
  # when initiating many parallel requests marshaling the certificate store is
  # the most likely point of failure (see issue #382). Any program attempting
  # to leverage this pseudo safety should make a call to this method (i.e.
  # `Stripe.ca_store`) in their initialization code because it marshals lazily
  # and is itself not thread safe.
  def self.ca_store
    @ca_store ||= begin
      store = OpenSSL::X509::Store.new
      store.add_file(ca_bundle_path)
      store
    end
  end

  # map to the same values as the standard library's logger
  LEVEL_DEBUG = Logger::DEBUG
  LEVEL_ERROR = Logger::ERROR
  LEVEL_INFO = Logger::INFO

  # When set prompts the library to log some extra information to $stdout and
  # $stderr about what it's doing. For example, it'll produce information about
  # requests, responses, and errors that are received. Valid log levels are
  # `debug` and `info`, with `debug` being a little more verbose in places.
  #
  # Use of this configuration is only useful when `.logger` is _not_ set. When
  # it is, the decision what levels to print is entirely deferred to the logger.
  def self.log_level
    @log_level
  end

  def self.log_level=(val)
    # Backwards compatibility for values that we briefly allowed
    if val == "debug"
      val = LEVEL_DEBUG
    elsif val == "info"
      val = LEVEL_INFO
    end

    if !val.nil? && ![LEVEL_DEBUG, LEVEL_ERROR, LEVEL_INFO].include?(val)
      raise ArgumentError, "log_level should only be set to `nil`, `debug` or `info`"
    end
    @log_level = val
  end

  # Sets a logger to which logging output will be sent. The logger should
  # support the same interface as the `Logger` class that's part of Ruby's
  # standard library (hint, anything in `Rails.logger` will likely be
  # suitable).
  #
  # If `.logger` is set, the value of `.log_level` is ignored. The decision on
  # what levels to print is entirely deferred to the logger.
  def self.logger
    @logger
  end

  def self.logger=(val)
    @logger = val
  end

  def self.max_network_retries
    @max_network_retries
  end

  def self.max_network_retries=(val)
    @max_network_retries = val.to_i
  end

  # Sets some basic information about the running application that's sent along
  # with API requests. Useful for plugin authors to identify their plugin when
  # communicating with Stripe.
  #
  # Takes a name and optional version and plugin URL.
  def self.set_app_info(name, version: nil, url: nil)
    @app_info = {
      name: name,
      url: url,
      version: version,
    }
  end

  # DEPRECATED. Use `Util#encode_parameters` instead.
  def self.uri_encode(params)
    Util.encode_parameters(params)
  end
  private_class_method :uri_encode
  class << self
    extend Gem::Deprecate
    deprecate :uri_encode, "Stripe::Util#encode_parameters", 2016, 1
  end
end

Stripe.log_level = ENV["STRIPE_LOG"] unless ENV["STRIPE_LOG"].nil?
